### PR TITLE
add edge classes

### PR DESCRIPTION
**Pull request by chris-soelistyo**
- Generalised edge classification such that more than two classes are possible
- Uses simple cross entropy rather than the custom function, since the output is now comparable to node classification
- Only difference is `node_x.size() == (batch, num_node_classes)` while `edge_x.size() == (batch, num_edges, num_edge_classes)` for each subgraph.

### CODE REVIEW

File: grace/models/classifier.py [hunk 1]
```diff
@@ -26,14 +26,16 @@ def __init__(
...
+        edge_output_classes: int = 2,
     ):
         super(GCN, self).__init__()
         torch.manual_seed(12345)
         self.conv1 = GCNConv(input_channels, hidden_channels)
         self.conv2 = GCNConv(hidden_channels, hidden_channels)
         self.conv3 = GCNConv(hidden_channels, hidden_channels)
-        self.linear = Linear(hidden_channels, output_classes)
+        self.node_classifier = Linear(hidden_channels, node_output_classes)
+        self.edge_classifier = Linear(hidden_channels * 2, edge_output_classes)
```
Comment (quantumjot) [inline grace/models/classifier.py:38]: This is because you cat the source and target node embeddings?

Comment (quantumjot) [reply grace/models/classifier.py:38]: Maybe add a comment here to describe the logic. Better still, add a Notes section to the docstrings

File: grace/models/classifier.py [hunk 2]
```diff
@@ -26,14 +26,16 @@ def __init__(
         input_channels: int,
         hidden_channels: int,
         *,
-        output_classes: int = 2,
+        node_output_classes: int = 2,
+        edge_output_classes: int = 2,
     ):
         super(GCN, self).__init__()
         torch.manual_seed(12345)
```
Comment (quantumjot) [inline grace/models/classifier.py:33]: Is this needed here?

File: grace/models/classifier.py [hunk 3]
```diff
@@ -50,5 +52,16 @@ def forward(
...
-        return x, embeddings
+        node_x = self.node_classifier(x)
+
+        src, dst = edge_index
+        edge_features = torch.cat(
+            [embeddings[..., src, :], embeddings[..., dst, :]], axis=-1
+        )
+
+        if edge_features.ndim == 2:
+            edge_features = edge_features.unsqueeze(0)
```
Comment (quantumjot) [inline grace/models/classifier.py:63]: For the case where there is no batch?

File: grace/models/train.py [hunk 4]
```diff
@@ -49,18 +34,19 @@ def train_model(
...
 
         for data in train_loader:
-            out_x, out_embedding = model(data.x, data.edge_index, data.batch)
-            loss_node = node_criterion(out_x, data.y)
-            loss_edge = edge_criterion(
-                out_embedding, data.edge_label, data.edge_index, masked_class
-            )
+            node_x, edge_x = model(data.x, data.edge_index, data.batch)
+            loss_node = node_criterion(node_x, data.y)
+            loss_edge = edge_criterion(edge_x, data.edge_label)
```
Comment (quantumjot) [inline grace/models/train.py:49]: Nice and clean - I like it!

**Review by quantumjot — APPROVED** [review]
Looks good to me.

File: grace/models/classifier.py (hunk 3 above)
Comment (chris-soelistyo) [reply grace/models/classifier.py:63]: Yup!

File: grace/models/classifier.py (hunk 2 above)
Comment (quantumjot) [reply grace/models/classifier.py:33]: Just incase it's not clear - I meant the seed

Comment (chris-soelistyo) [reply grace/models/classifier.py:40]: Ah good spot! No, not needed I don't think.
